internal/datanode/segment_sync_policy.go

```diff
@@ -17,25 +17,43 @@
 package datanode
 
 import (
+	"sort"
 	"time"
 
-	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/util/hardware"
 	"github.com/milvus-io/milvus/internal/util/tsoutil"
-	"go.uber.org/zap"
 )
 
-// segmentSyncPolicy sync policy applies to segment
-type segmentSyncPolicy func(segment *Segment, ts Timestamp) bool
+// segmentsSyncPolicy sync policy applies to segments
+type segmentSyncPolicy func(segments []*Segment, ts Timestamp) []*Segment
 
-// syncPeriodically get segmentSyncPolicy with segment sync periodically.
+// syncPeriodically get segmentSyncPolicy with segments sync periodically.
 func syncPeriodically() segmentSyncPolicy {
-	return func(segment *Segment, ts Timestamp) bool {
-		endTime := tsoutil.PhysicalTime(ts)
-		lastSyncTime := tsoutil.PhysicalTime(segment.lastSyncTs)
-		shouldSync := endTime.Sub(lastSyncTime) >= Params.DataNodeCfg.SyncPeriod.GetAsDuration(time.Second) && !segment.isBufferEmpty()
-		if shouldSync {
-			log.Info("sync segment periodically ", zap.Time("now", endTime), zap.Time("last sync", lastSyncTime))
+	return func(segments []*Segment, ts Timestamp) []*Segment {
+		segsToSync := make([]*Segment, 0)
+		for _, seg := range segments {
+			endTime := tsoutil.PhysicalTime(ts)
+			lastSyncTime := tsoutil.PhysicalTime(seg.lastSyncTs)
+			shouldSync := endTime.Sub(lastSyncTime) >= Params.DataNodeCfg.SyncPeriod.GetAsDuration(time.Second) && !seg.isBufferEmpty()
+			if shouldSync {
+				segsToSync = append(segsToSync, seg)
+			}
 		}
-		return shouldSync
+		return segsToSync
+	}
+}
+
+// syncMemoryTooHigh get segmentSyncPolicy with segment sync periodically.
+func syncMemoryTooHigh() segmentSyncPolicy {
+	return func(segments []*Segment, ts Timestamp) []*Segment {
+		if Params.DataNodeCfg.MemoryForceSyncEnable.GetAsBool() &&
+			hardware.GetMemoryUseRatio() >= Params.DataNodeCfg.MemoryForceSyncThreshold.GetAsFloat() {
+			toSyncSegmentNum := int(Params.DataNodeCfg.MemoryForceSyncSegmentRatio.GetAsFloat() * float64(len(segments)))
+			sort.Slice(segments, func(i, j int) bool {
+				return segments[i].numRows > segments[j].numRows
+			})
+			return segments[:toSyncSegmentNum]
```

Comment:
maybe we only sync the segment with largest memory?

because under high watermark we want to release some memory ASAP.

Also add a log there so we know why the sync happened